### PR TITLE
Fix cross-attention head mask for Torch encoder-decoder models

### LINKED ISSUE #10540: 🐛 Bug in attention head mask for cross-attention module in encoder-decoder models
Currently, encoder-decoder models use either `head_mask` or `decoder_head_mask` for masking attention heads in cross-attention modules. Both cases are not perfectly correct. Furthermore, MHA in cross-attention modules shares the parameters with the decoder, i.e. `shape = (decoder.num_layers, decoder.num_attention_heads)`, therefore, the usage of encoder `head_mask` in the cross-attention module may lead to errors due to the shape mismatch.

<hr>

**My contribution:** I will take care of this issue this weekend.

<hr>

**Reviewers:** @patil-suraj @patrickvonplaten

### PR DESCRIPTION
**Pull request by stancld**
1. This PR fixes head masking for the cross-attention module in the following models:

- BART,
- Blenderbot,
- Blenderbot_small,
- FSMT,
- LED,
- M2M_100,
- Marian,
- MBart,
- Pegasus.
- T5

2. This PR also contains slight changes in docstrings so that it will be clear that `head_mask` is related to the config of an encoder and the shape of `decoder_head_mask` and `cross_head_mask` depends on the config of a decoder.
 
3. This PR enables `test_headmasking` for M2M_100 model.

<hr>

**Reviewers:** @patrickvonplaten @patil-suraj 

<hr>

Fixes: #10540

### CODE REVIEW

**Comment by stancld** [comment]
I'll include T5, FSMT and LED models into this PR as well to resolve the issue with a  forward signature (and without adding any other `if` condition which would be removed immediately afterwards).

**Comment by stancld** [comment]
Hi @patrickvonplaten & @patil-suraj, my PR does not pass one test, however, I am not able to reproduce this error on my local (I can't even find the file at `src/transformers/models/new_enc_dec/modeling_new_enc_dec.py` in the repo, which is the one where should be a problem with a copy inconsistency)

**Comment by LysandreJik** [comment]
Hi @stancld thank you for your work! The issue is because you have updated all model files (thank you!!), but you haven't updated the template. The template is used when adding a new model, it's available [here](https://github.com/huggingface/transformers/tree/master/templates/adding_a_new_model).

For example, these lines should probably be updated to include the `cross_head_mask`:

https://github.com/huggingface/transformers/blob/44f64132a5f50726f9de4467ed745421c3b11ab3/templates/adding_a_new_model/cookiecutter-template-%7B%7Bcookiecutter.modelname%7D%7D/modeling_%7B%7Bcookiecutter.lowercase_modelname%7D%7D.py#L2070-L2085

**Comment by stancld** [comment]
@LysandreJik - Thank you very much for the clarification :)

**Comment by LysandreJik** [comment]
Wonderful, the tests pass! Thanks for handling the templates.

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -368,7 +368,7 @@ def forward(
         encoder_hidden_states: Optional[torch.Tensor] = None,
         encoder_attention_mask: Optional[torch.Tensor] = None,
         layer_head_mask: Optional[torch.Tensor] = None,
-        encoder_layer_head_mask: Optional[torch.Tensor] = None,
+        cross_layer_head_mask: Optional[torch.Tensor] = None,
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:371]:
```suggestion
        cross_attn_layer_head_mask: Optional[torch.Tensor] = None,
```

File: src/transformers/models/bart/modeling_bart.py [hunk 2]
```diff
@@ -382,9 +382,9 @@ def forward(
             encoder_attention_mask (:obj:`torch.FloatTensor`): encoder attention mask of size
                 `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
             layer_head_mask (:obj:`torch.FloatTensor`): mask for attention heads in a given layer of size
-                `(config.encoder_attention_heads,)`.
-            encoder_layer_head_mask (:obj:`torch.FloatTensor`): mask for encoder attention heads in a given layer of
-                size `(config.encoder_attention_heads,)`.
+                `(encoder_attention_heads,)`.
+            cross_layer_head_mask (:obj:`torch.FloatTensor`): mask for cross-attention heads in a given layer of
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:386]:
```suggestion
            cross_attn_layer_head_mask (:obj:`torch.FloatTensor`): mask for cross-attention heads in a given layer of
```

File: src/transformers/models/bart/modeling_bart.py [hunk 3]
```diff
@@ -382,9 +382,9 @@ def forward(
...
+                size `(decoder_attention_heads,)`.
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:387]:
```suggestion
                size ``(decoder_attention_heads,)``.
```

File: src/transformers/models/bart/modeling_bart.py [hunk 4]
```diff
@@ -419,7 +419,7 @@ def forward(
                 hidden_states=hidden_states,
                 key_value_states=encoder_hidden_states,
                 attention_mask=encoder_attention_mask,
-                layer_head_mask=encoder_layer_head_mask,
+                layer_head_mask=cross_layer_head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:422]:
```suggestion
                layer_head_mask= cross_attn_layer_head_mask,
```

File: src/transformers/models/bart/modeling_bart.py [hunk 5]
```diff
@@ -597,18 +597,24 @@ def __init_subclass__(self):
...
+            - 0 indicates the head is **masked**.
 
-        decoder_head_mask (:obj:`torch.Tensor` of shape :obj:`(num_layers, num_heads)`, `optional`):
+        decoder_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
             Mask to nullify selected heads of the attention modules in the decoder. Mask values selected in ``[0, 1]``:
 
             - 1 indicates the head is **not masked**,
             - 0 indicates the head is **masked**.
 
+        cross_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:612]:
```suggestion
        cross_attn_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
```

File: src/transformers/models/bart/modeling_bart.py [hunk 6]
```diff
@@ -874,7 +880,7 @@ def forward(
         encoder_hidden_states=None,
         encoder_attention_mask=None,
         head_mask=None,
-        encoder_head_mask=None,
+        cross_head_mask=None,
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:883]:
```suggestion
        cross_attn_head_mask =None,
```

File: src/transformers/models/bart/modeling_bart.py [hunk 7]
```diff
@@ -911,18 +917,18 @@ def forward(
...
-            head_mask (:obj:`torch.Tensor` of shape :obj:`(num_layers, num_heads)`, `optional`):
+            head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
                 Mask to nullify selected heads of the attention modules. Mask values selected in ``[0, 1]``:
 
                 - 1 indicates the head is **not masked**,
-                - 0 indicates the heas is **masked**.
+                - 0 indicates the head is **masked**.
 
-            encoder_head_mask (:obj:`torch.Tensor` of shape :obj:`(num_layers, num_heads)`, `optional`):
+            cross_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:926]:
```suggestion
            cross_attn_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
```

File: src/transformers/models/bart/modeling_bart.py [hunk 8]
```diff
@@ -992,11 +998,12 @@ def forward(
         all_cross_attentions = () if (output_attentions and encoder_hidden_states is not None) else None
         next_decoder_cache = () if use_cache else None
 
-        # check if head_mask has a correct number of layers specified if desired
-        if head_mask is not None:
-            assert head_mask.size()[0] == (
-                len(self.layers)
-            ), f"The head_mask should be specified for {len(self.layers)} layers, but it is for {head_mask.size()[0]}."
+        # check if head_mask/cross_head_mask has a correct number of layers specified if desired
+        for attn_mask, mask_name in zip([head_mask, cross_head_mask], ["head_mask", "cross_head_mask"]):
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:1002]:
```suggestion
        for attn_mask, mask_name in zip([head_mask, cross_attn_head_mask], ["head_mask", "cross_head_mask"]):
```

**Review by patil-suraj — COMMENTED** [review]
This looks good to me. Thanks a lot for fixing this!

I would prefer slightly explicit names

`cross_layer_head_mask` => `cross_attn_layer_head_mask`

and `cross_head_mask` to `cross_attn_head_mask`.

**Comment by stancld** [comment]
@patil-suraj Thank you for your review and the suggestions. I agree with the change in variable naming. The new one is more accurate. I will change it everywhere accordingly :)

**Comment by stancld** [comment]
Hi @patrickvonplaten, I've just rebased this branch to the current master to keep this PR up to date. Could you please review this one? Thanks a lot! :)

**Comment by LysandreJik** [comment]
Hey @stancld, Patrick is off for a couple of weeks but will take a look at this as soon as he's back :)

File: src/transformers/models/bart/modeling_bart.py [hunk 9]
```diff
@@ -912,18 +918,18 @@ def forward(
...
                 Mask to nullify selected heads of the attention modules. Mask values selected in ``[0, 1]``:
 
                 - 1 indicates the head is **not masked**,
-                - 0 indicates the heas is **masked**.
+                - 0 indicates the head is **masked**.
 
-            encoder_head_mask (:obj:`torch.Tensor` of shape :obj:`(num_layers, num_heads)`, `optional`):
+            cross_attn_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
                 Mask to nullify selected heads of the attention modules in encoder to avoid performing cross-attention
                 on hidden heads. Mask values selected in ``[0, 1]``:
```
Comment (patil-suraj) [inline src/transformers/models/bart/modeling_bart.py:929]:
```suggestion
            cross_attn_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
                Mask to nullify selected heads of the cross-attention modules in decoder to avoid performing cross-attention
                on encoder hidden heads. Mask values selected in ``[0, 1]``:
```

File: src/transformers/models/blenderbot_small/modeling_blenderbot_small.py [hunk 10]
```diff
@@ -875,18 +881,18 @@ def forward(
...
                 Mask to nullify selected heads of the attention modules. Mask values selected in ``[0, 1]``:
 
                 - 1 indicates the head is **not masked**,
-                - 0 indicates the heas is **masked**.
+                - 0 indicates the head is **masked**.
 
-            encoder_head_mask (:obj:`torch.Tensor` of shape :obj:`(num_layers, num_heads)`, `optional`):
+            cross_attn_head_mask (:obj:`torch.Tensor` of shape :obj:`(decoder_layers, decoder_attention_heads)`, `optional`):
                 Mask to nullify selected heads of the attention modules in encoder to avoid performing cross-attention
                 on hidden heads. Mask values selected in ``[0, 1]``:
```
Comment (patil-suraj) [inline src/transformers/models/blenderbot_small/modeling_blenderbot_small.py:892]: same comment as above

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for taking care of this @stancld, LGTM!

If you fix the docstring `cross_attn_head_mask` in few files, it should be good to merge.

**Comment by patrickvonplaten** [comment]
The PR looks good to me :-) 

Think we just need to fix the docstring. @stancld - let me know if you need help regarding the docstring

**Comment by stancld** [comment]
@patrickvonplaten The docstring should be fixed now. I forgot a line from a conflict there.. :)
